De-sync of ports in older versions of BISDN linux (#103)
bisdn/bisdn.github.io: limitations.md

## Missing routes for EIGRP with flapping ports

As documented in the currently open upstream FRR issue [#7299](https://github.com/FRRouting/frr/issues/7299), some routes may get dropped or are not correctly received when ports are flapping during EIGRP session establishment. For now, we recommend the workaround of restarting FRR after all ports are up if this behavior is observed.

> ## Port state sync issue on boot caused by OF-DPA

Reviewer: What is the result of this? How can a customer identify if he is running into this issue?

Author: Commented suggestion below.

Reviewer 2: Maybe something along 
```suggestion
## Ports connected during boot may sometimes show as having no carrier in Linux
```
makes it clearer about the visible symptoms (and situation where it happens)

Author: Updated the description of the problem and how to resolve it

All releases of BISDN Linux prior to version 4.0 suffer from an issue where
the port state might end up out of sync

This is caused by a race in OF-DPA, where OF-DPA first initializes ports with
their current state, and only then registers the linkscan handler, which
is responsible for updating OF-DPA's port state. This creates a window where
OF-DPA will miss any physical link state changes happening.

Any port state changes happening between the initial read out and the

> successful registration of the handler will be missed.

Author: ```suggestion
successful registration of the handler will be missed.

The port sync issue may be identified by the inability to set a port up, even though there is a physical cable plugged into the port.

E.g having a cable into port2, and running ``ip link set port2 up``, then the output of ``ip link show port2`` should show the state as ``UP``. If the state is ``DOWN``. you likely have a port state sync issue in OF-DPA, which you can resolve by first disabling and then enabling the port again.

``client_drivshell port 2 Enable=false``
``client_drivshell port 2 Enable=true``
```

## Missing routes for EIGRP with flapping ports

As documented in the currently open upstream FRR issue [#7299](https://github.com/FRRouting/frr/issues/7299), some routes may get dropped or are not correctly received when ports are flapping during EIGRP session establishment. For now, we recommend the workaround of restarting FRR after all ports are up if this behavior is observed.

## Port state sync issue on boot caused by OF-DPA

All releases of BISDN Linux prior to version 4.0 suffer from an issue where
port state might end up out of sync if network configuration is applied during,

> or shortly after, booting up, e.g. using systemd-networkd.

Author: added

> or shortly after, booting up, e.g. using systemd-networkd.

Reviewer 2: This has nothing to do with systemd-networkd - this is purely about the physical state, i.e. if a cable is plugged in and has a link. The issue then can happen regardless if the port is administratively up or do.

> or shortly after, booting up, e.g. using systemd-networkd.

Reviewer 3: systemd-networkd is simply an example of how "network configuration is applied during or shortly after, booting up", as opposed to logging in some time later and configuring the network manually, e.g. using ip(8). So if the issue has nothing to do with systemd-networkd, then "if network configuration is applied during, or shortly after, booting up" is also wrong.

> or shortly after, booting up, e.g. using systemd-networkd.

Reviewer 2: Right, this part is also wrong.

The issue happens if ports are physically connected, and the link gets established during that window. Then the link up event will be missed, and the internal cached link state will remain in the wrong state (until the port is brought down and up again). This all happens before OF-DPA/ofagent even accepts connections, so before any port interfaces exist in linux.

> or shortly after, booting up, e.g. using systemd-networkd.

Author: So the 'link up event' here has nothing to do with configuring links/ports using sytemd or iproute2. How should we reword it so that it is correct?

"If changes in port state during this time?"
I really can't think of how to word this properly.

> or shortly after, booting up, e.g. using systemd-networkd.

Author: Reworded quite a bit, can does it look better now @KanjiMonster?

This is caused by a race in OF-DPA, where OF-DPA first initializes ports with

> their current state, and only then registers notifications about port state

Author: I am not entirely happy with this: "initializes ports with their current state", what exactly does current state mean here @KanjiMonster

Author: Or can we be more precise in what we mean.

Author: Now mentions the linkscan handler, which is responsible for the port state updates in OF-DPA

changes.

Any port state changes happening between the initial read out and the
successful registration of the handler will be missed.

> 

Author: fixed

## Missing routes for EIGRP with flapping ports

As documented in the currently open upstream FRR issue [#7299](https://github.com/FRRouting/frr/issues/7299), some routes may get dropped or are not correctly received when ports are flapping during EIGRP session establishment. For now, we recommend the workaround of restarting FRR after all ports are up if this behavior is observed.

## Ports connected during boot may sometimes show as having no carrier in Linux

> All releases of BISDN Linux prior to version 4.0 suffer from an issue where

Reviewer 2: ```suggestion
All releases of BISDN Linux prior to version 3.7.3 suffer from an issue where
```

> the port state might end up out of sync

Reviewer 2: ```suggestion
the port state might end up out of sync.
```

This is caused by a race in OF-DPA, where OF-DPA first initializes ports with
their current state, and only then registers the linkscan handler, which
is responsible for updating OF-DPA's port state. This creates a window where
OF-DPA will miss any physical link state changes happening.

Any port state changes happening between the initial read out and the
successful registration of the handler will be missed.

> The port sync issue may be identified by the link inability to set a port up

Reviewer 2: ```suggestion
The port sync issue may be identified by the link's inability to set a port up
```
